src/code/fault_n64.c

```diff
@@ -0,0 +1,845 @@
+    "Virtual coherency on data",
+};
+
+const char* sFpExceptionNames[] = {
+    "Unimplemented operation", "Invalid operation", "Division by zero", "Overflow", "Underflow", "Inexact operation",
+};
+
+u16 sFaultFontColor = GPACK_RGBA5551(255, 255, 255, 1);
+
+Input sFaultInputs[MAXCONTROLLERS];
+
+FaultMgr gFaultMgr;
+
+STACK(sFaultStack, 0x400);
+StackEntry sFaultStackInfo;
+FaultCursorCoords sFaultCursorPos;
+
+vs32 sFaultExit;
+vs32 gFaultMsgId;
+vs32 sFaultDisplayEnable;
+OSThread* sFaultFaultedThread;
+s32 B_80122570[0x10];
+
+void Fault_SleepImpl(u32 ms) {
+    Sleep_Msec(ms);
+}
+
+void Fault_WaitForInputImpl(void) {
+    Input* inputs = sFaultInputs;
+    u16 btnPress;
+
+    do {
+        Fault_SleepImpl(0x10);
+        PadMgr_RequestPadData(&gPadMgr, inputs, 0);
+        btnPress = inputs[0].press.button;
+    } while (!CHECK_BTN_ANY(btnPress, (BTN_A | BTN_B | BTN_START | BTN_CRIGHT | BTN_CLEFT | BTN_CDOWN | BTN_CUP)));
+}
+
+void Fault_WaitForInput(void) {
+    Fault_WaitForInputImpl();
+}
+
+void Fault_DrawRec(s32 x, s32 y, s32 w, s32 h, u16 color) {
+    s32 i;
+    s32 j;
+    u16* fbPtr;
+
+    fbPtr = gFaultMgr.fb + (gFaultMgr.fbWidth * y) + x;
+
+    for (i = 0; i < h; i++) {
+        j = 0;
+        for (j = 0; j < w; j++) {
+            *fbPtr = color;
+            fbPtr++;
+        }
+        fbPtr += gFaultMgr.fbWidth - w;
+    }
+    osWritebackDCacheAll();
+}
+
+void Fault_DrawRecBlack(s32 x, s32 y, s32 w, s32 h) {
+    Fault_DrawRec(x, y, w, h, GPACK_RGBA5551(0, 0, 0, 1));
+}
+
+void Fault_DrawCharImpl(s32 x, s32 y, char c) {
+    u32* dataPtr;
+    s32 shift = c % 4;
+    s32 i;
+    s32 j;
+    u16* fbPtr;
+    u32 data;
+
+    dataPtr = (u32*)sFaultDrawerFont + ((c / 8 * 0x10) + ((c & 4) >> 2));
+    fbPtr = gFaultMgr.fb + (gFaultMgr.fbWidth * y) + x;
+
+    for (i = 0; i < 8; i++) {
+        u32 mask;
+
+        mask = (1 << 28) << shift;
+        data = *dataPtr;
+        for (j = 0; j < 8; j++) {
+            if (mask & data) {
+                *fbPtr = sFaultFontColor;
+            }
+            fbPtr++;
+            mask >>= 4;
+        }
+        dataPtr += 2;
+        fbPtr += gFaultMgr.fbWidth - 8;
+    }
+    osWritebackDCacheAll();
+}
+
+void Fault_DrawChar(s32 x, s32 y, u8 c) {
+    Fault_DrawCharImpl(x, y, c);
+}
+
+void Fault_DrawCornerRec(u16 color) {
+    Fault_DrawRec(22, 16, 8, 1, color);
+}
+
+void Fault_DrawCornerRecRed(void) {
+    Fault_DrawCornerRec(GPACK_RGBA5551(255, 0, 0, 1));
+}
+
+void Fault_DrawCornerRecYellow(void) {
+    Fault_DrawCornerRec(GPACK_RGBA5551(255, 255, 0, 1));
+}
+
+void func_800AE05C(void) {
+}
+
+void* Fault_PrintCallbackDraw(void* arg, const char* str, size_t len) {
+    FaultCursorCoords* coords = arg;
+
+    for (; len != 0; len--, str++) {
+        if (*str == '\n') {
+            coords->x = 320;
+        } else {
+            Fault_DrawChar(coords->x, coords->y, *str);
+            coords->x += 6;
+        }
+        if (coords->x >= 277) {
```

Comment:
```suggestion
        if (coords->x > 276) {
```
Kind of nitpicky but 276 does seem to have more signifcance based on the `Fault_DrawRecBlack` call.